### PR TITLE
[Feature] PersonaPlex (Moshi-based full-duplex S2S): native vLLM port + duplex serving

### PR DESCRIPTION
**Pull request by linyueqian**
## Summary

Adds **PersonaPlex** (`nvidia/personaplex-7b-v1`, a Moshi-based full-duplex speech-to-speech model) to vLLM-Omni: a native vLLM port of the model plus a full-duplex serving path that runs the **official PersonaPlex web client** against our own engine.

This is **experimental** — the duplex serving framework lives under `vllm_omni/experimental/fullduplex/`. The model itself is a normal model port under `vllm_omni/model_executor/models/personaplex/`.

## What's here

**Native model port** (`model_executor/models/personaplex/`)
- Helium temporal transformer, depformer (RQ code predictor), input embeddings (`embed_codes`), and Code2Wav (Mimi) — each parity-checked against the Moshi reference.
- Staged talker + code2wav pipeline on the omni engine (paged-KV), with the acoustic de-delay applied to agent codes before Mimi.

**Full-duplex serving** (`experimental/fullduplex/personaplex/`)
- A `FrameStepper`/`PersonaPlexSession` lockstep driver (one 80 ms user frame in → one agent frame + text out).
- A WebSocket server that **serves the official PersonaPlex web client** (the same `dist.tgz` bundle `moshi.server` uses) at `/` and implements its protocol at `/api/chat` (Opus via `sphn`, tags `\x00` ready / `\x01` audio / `\x02` text), driven by the native components. A raw-PCM endpoint (`/v1/audio/duplex`) is kept for Opus-free tests.

The only changes outside the PersonaPlex/experimental trees are 4 registration hooks (model + pipeline registries, `arg_utils`, `tts_adapters/__init__`).

## Verification

- **Coherent e2e speech** (offline staged pipeline and the duplex WS path), ASR-checked — e.g. *"You want fluffy rice, so first use a pot with a lid. Bring water to a boil, add a splash of oil, then add the rice, let it simmer…"*.
- **Real-time:** native per-frame step **RTF ≈ 0.33** (mean ~26 ms vs the 80 ms budget, 0 over budget) after CUDA-graphing the native path the same way Moshi graphs its own forwards. Server-side `/api/chat` delivery is a steady ~82 ms/packet.
- The native Helium temporal matches Moshi at **cos ≈ 0.999** (teacher-forced).

## How to run

```bash
HF_TOKEN=... CUDA_VISIBLE_DEVICES=0 bash examples/online_serving/personaplex/run_server.sh
# open http://localhost:8124/  -> the official PersonaPlex web client (mic in the browser)
```

Headless client + protocol docs: `examples/online_serving/personaplex/` (`duplex_client.py`, `README.md`). Offline inference: `examples/offline_inference/personaplex/personaplex_offline.py`. Unit tests: `tests/fullduplex/`.

## Update: concurrent serving (elastic batching)

`--batch-size N` now serves **N concurrent duplex conversations on one engine** (default `1` keeps the original single-session path untouched). Measured on a single 140 GB-class GPU (greedy, native components):

| B | ms/tick | RTF/stream | VRAM |
|---|---------|-----------|------|
| 2 | 28.0 | 0.35 | 25.1 GiB |
| 8 | 32.6 | 0.41 | 42.0 GiB |
| 16 | 45.0 | 0.56 | 64.8 GiB |
| 32 | 70.2 | 0.88 | 110.5 GiB |

i.e. up to **32 realtime conversations per GPU** within the 80 ms frame budget (vs. 1 before).

**How it works.** All slots advance in strict lockstep on a shared wall-clock 12.5 Hz stepping thread; idle slots are fed silence so a caller can join at any tick. A finished slot is **recycled** for a new caller by masking its KV window (per-row `start_offset` on the ring cache -- writes stay uniform, so no scatter and no paging) and replaying its system prompt per-row (stored voice embeddings + persona tokens via `force_rows`) while the other slots keep talking. The moshi changes ship as runtime patches (`elastic.py`, no dependency fork); the per-slot recycle design is ported from [Hcoder10/personaplex-business-ft](https://github.com/Hcoder10/personaplex-business-ft) `cb-fix` (MIT), with correctness fixes found during verification (Mimi-vs-LM reset semantics; a split embed/temporal graph path so mixed prefill ticks are bit-identical to normal ticks for live rows).

**E2E latency** (verified on a 141 GB Hopper-class GPU, eager, realtime budget = 80 ms/frame @ 12.5 Hz):

| Config | Per-tick latency | Realtime? |
| --- | --- | --- |
| `--batch-size 4` (4 concurrent conversations) | ~70-74 ms | yes, all four |
| single session | well under budget (headroom above B=4) | yes |

Current gates on this head: 56/56 GPU-free (`tests/fullduplex/`), 2/2 GPU e2e
(`tests/e2e/fullduplex/test_personaplex_elastic.py`: bit-exact recycle isolation +
recycled-slot coherence against a real spoken question).

**Verification** (single GPU, greedy decoding makes any cross-slot leak a bit difference):
- **Isolation, bit-exact:** recycling one slot mid-flight leaves every other slot's audio bit-identical to a run that never recycled (`tests/e2e/fullduplex/test_personaplex_elastic.py`).
- **Replay exactness:** a recycled slot's opening live ticks match a pristine boot at numerical-noise level (~1e-5).
- **Coherence:** a recycled slot fed a real spoken question replies coherently in its inner-monologue stream.
- **Server smoke:** 5 concurrent WS clients vs `--batch-size 4`: four held ~15 s realtime streams (incl. three overlapping per-slot recycle prefills), the fifth was rejected cleanly; released slots were reused.
- 56 GPU-free unit tests (`tests/fullduplex/`), including the slot-manager suite, the
  lease-drain regression tests, and the model-owned lockstep-runtime contract tests.

Batched-mode limitation: the voice prompt is the fixed server default (`.pt` embeddings); per-connection `voice_prompt` is ignored with a warning. Per-slot **personas** are fully supported (replayed on recycle).

## Known limitations / follow-ups

- **Greedy decoding by default.** The temporal is cos-0.999, not bit-exact, so sampling diverges; greedy is reliably coherent. Bit-exact temporal (to enable sampling) is the main follow-up.
- Real-time smoothness for **remote** browsers is network-bound (the official client buffers for localhost); run the client close to the server for best results.

---

Marked experimental; happy to split into smaller PRs if preferred for review.

### CODE REVIEW

File: vllm_omni/engine/arg_utils.py [hunk 1]
```diff
@@ -74,6 +77,7 @@ def _register_omni_hf_configs() -> None:
         ("moss_tts_local", MossTTSLocalConfig),
         ("moss_tts_realtime", MossTTSRealtimeConfig),
         ("qwen3_tts", Qwen3TTSConfig),
+        ("personaplex", PersonaPlexConfig),
```
Comment (hsliuustc0106) [inline vllm_omni/engine/arg_utils.py:85]: Registering the config class is not enough for this repo, because PersonaPlex config.json is empty. Add the PersonaPlex architectures to `_ARCH_TO_MODEL_TYPE`; otherwise `create_model_config()` never patches `model_type=personaplex`, and the staged pipeline cannot load the HF config without manual overrides.

File: vllm_omni/model_executor/stage_input_processors/personaplex.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    # Codes live in the server-side request's additional_information under
+    # ("codes","audio") (talker_mtp_output_key), not in multimodal_output (latent).
+    def _codes_from(src: Any) -> torch.Tensor | None:
+        if isinstance(src, dict):
+            nested = src.get("codes")
+            a = nested.get("audio") if isinstance(nested, dict) else None
+            return a if a is not None else src.get("codes.audio")
+        return None
+
+    audio = _codes_from(getattr(request, "additional_information", None)) or _codes_from(multimodal_output)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_input_processors/personaplex.py]: This will raise when the first branch returns a Tensor: Python tries to evaluate its truth value for `or`. Use an explicit `if audio is None` fallback like the full-payload path.

File: vllm_omni/model_executor/stage_input_processors/personaplex.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    cfg = raw_cfg.get("extra", raw_cfg) if isinstance(raw_cfg, dict) else {}
+    chunk = int(cfg.get("codec_chunk_frames", 25))
+
+    if len(frames) < chunk and not finished:
+        return None
+    if not frames:
+        return _empty_finished_payload() if finished else None
+
+    stacked = torch.stack(frames, dim=0)  # [F, dep_q]
+    buf[request_id] = []
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_input_processors/personaplex.py]: This drops the boundary frame needed for de-delay. `_agent_codes_to_codebook_major()` builds cb1..7 from the next frame, so after emitting a non-final chunk the last frame must stay in the buffer as carry-over; otherwise one acoustic frame is lost/misaligned per chunk.

File: vllm_omni/entrypoints/openai/tts_adapters/personaplex.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        # Route through the shared dispatcher (which forwards to the
+        # PersonaPlex-specific validator for this model type) rather than a leaf
+        # validator directly, matching the Qwen3-TTS path.
+        return self.ctx.server._validate_tts_request(request)
+
+    async def build(
+        self, request: "OpenAICreateSpeechRequest", sampling_params_list: list, has_inline_ref_audio: bool
+    ) -> PreparedRequest:
+        # built by lead: server-side prompt/param builder for PersonaPlex.
+        prompt, tts_params, warmup_key = await self.ctx.server._build_personaplex_request(request)
```
Comment (hsliuustc0106) [inline vllm_omni/entrypoints/openai/tts_adapters/personaplex.py]: `_build_personaplex_request` is not defined on `OmniOpenAIServingSpeech`, and `serving_speech.py` also never detects `model_stage="personaplex"` as TTS. This adapter is currently unreachable/incomplete, and will crash if detection is added without the builder.

File: vllm_omni/experimental/fullduplex/personaplex/serving/server.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+    # ---- official Moshi protocol (Opus over aiohttp WS) ---------------------
+
+    async def handle_chat(self, request: web.Request) -> web.WebSocketResponse:
+        ws = web.WebSocketResponse()
+        await ws.prepare(request)
+        persona = (request.query.get("text_prompt") or "").strip() or None
+        voice = (request.query.get("voice_prompt") or "").strip() or None
+        sr = self.engine.sample_rate
+        session = PersonaPlexSession(self.engine, self.config)
```
Comment (hsliuustc0106) [inline vllm_omni/experimental/fullduplex/personaplex/serving/server.py]: All WebSocket sessions share `self.engine`, and `session.open()` resets that engine streaming state. Since the docs state one conversation per server instance, please guard this with an active-session lock/reject path; a second client currently corrupts the first conversation.

File: vllm_omni/experimental/fullduplex/personaplex/serving/server.py [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+    return app
+
+
+def main() -> None:
+    ap = argparse.ArgumentParser()
+    ap.add_argument("--host", default="0.0.0.0")
+    ap.add_argument("--port", type=int, default=8124)
+    ap.add_argument("--persona", default=None)
+    ap.add_argument("--voice", default="NATF2.pt")
+    ap.add_argument("--greedy", action="store_true", default=True)
```
Comment (lishunyang12) [inline vllm_omni/experimental/fullduplex/personaplex/serving/server.py]: `--greedy` with `action="store_true", default=True` can never be set back to False — greedy is forced on. Drop the `default=True`.

File: vllm_omni/experimental/fullduplex/personaplex/config.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    """
+
+    hf_repo: str = "nvidia/personaplex-7b-v1"
+    voice_prompt: str = "NATF2.pt"
+    persona: str = DEFAULT_PERSONA
+
+    device: str = "cuda"
+    cpu_offload: bool = False
+
+    greedy: bool = False
```
Comment (lishunyang12) [inline vllm_omni/experimental/fullduplex/personaplex/config.py]: The native stepper hardcodes greedy `argmax` for both the text head and the depformer, so `greedy`/`temp_audio`/`temp_text`/`topk_*`/`seed` here are all dead. Wire them through or drop them so callers aren't misled.

File: vllm_omni/experimental/fullduplex/personaplex/serving/server.py [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+
+import numpy as np
+import sphn
+from aiohttp import WSMsgType, web
+from huggingface_hub import hf_hub_download
+
+from vllm_omni.experimental.fullduplex.personaplex.config import PersonaPlexConfig
+from vllm_omni.experimental.fullduplex.personaplex.engine import PersonaPlexEngine
+from vllm_omni.experimental.fullduplex.personaplex.native_stepper import (
+    NativePersonaPlexEngine as BatchedPersonaPlexEngine,
```
Comment (lishunyang12) [inline vllm_omni/experimental/fullduplex/personaplex/serving/server.py]: `BatchedPersonaPlexEngine` is just an alias for the same `NativePersonaPlexEngine` — the rename reads like a distinct engine. Import it under its real name.

**Review by lishunyang12 — COMMENTED** [review]
Greedy being forced on and the dead sampling knobs stood out; everything else is minor.

**Comment by Sy0307** [comment]
In `DuplexServer.handle_chat`, every connection shares the same `self.engine`, but `NativePersonaPlexEngine` has no internal locking — `open()` resets `_cache`/`_phys`/KV and `step()` mutates the same state frame by frame, so the `_active` lease is the only guard. The problem: when a client disconnects, if `proc_loop` is blocked in `await asyncio.to_thread(session.feed)` (server.py:164), `t.cancel(); await t` returns immediately while the worker thread keeps running `engine.step` (Python can't cancel a running thread). I verified this on the same Python 3.12 as prod: after cancel, the `await` returns at +0.000s with the thread still alive. Then `finally: self._active = False` (server.py:183) releases the lease right away. So a new connection that arrives within that ~1-tick (~75ms) window (a fast reconnect, or a client retrying after a 1013) acquires the lease and calls `open()` on the same engine from another thread, racing the still-running `step` — which can cause a CUDA illegal memory access (crashing the process) or silent cross-session state bleed. Consider mirroring the batched path (`threading.Lock` + epoch stale-guard + single tick thread): await the in-flight feed future before releasing the lease, or put a real lock around `open`/`feed`.

**Comment by Sy0307** [comment]
Four-commit update on top of the current unified full-duplex runtime:

- Integrated PersonaPlex with `/v1/realtime?duplex=1`, scheduler-backed Stage 0, and the unified audio/text data plane, providing a common serving and scheduling path.
- Hardened payload handoff, codec alignment, mixed batching, and request cleanup to prevent silent output, cross-request corruption, and state leaks.
- Added isolated multi-session Mimi codec slots and incremental batched decoding, enabling concurrency and slot reuse while reducing redundant decoding and memory allocations.
- Added unit and H20 E2E coverage for concurrent audible sessions, capacity rejection, slot reuse, input identity, and frame accounting; barge-in remains disabled until model state can be safely rolled back.

PTAK @linyueqian

**Review by Sy0307 — APPROVED** [review]
LGTM.

**Review by Peyman13l — APPROVED** [review]

**Review by Peyman13l — CHANGES_REQUESTED** [review]
123456
